[SDK-1920] Add Calling an API chapter to Angular quickstart

stevehobbsdev opened the pull request:
* Calling an API chapter added
* Updates 01-Login to use variables in snippets (workshop feedback)
* Updates 01-Login auth button snippet to use ngIf/Else construct (workshop feedback)

Commit:
Add chapter for calling APIs

Commit:
Fix snippets not using logged-in variables

Commit:
Refactor the auth button snippet to use if/else

vibronet commented on articles/quickstart/spa/angular-next/02-calling-an-api.md:
```diff
@@ -0,0 +1,186 @@
+---
+title: Call an API
+description: This tutorial demonstrates how to make API calls for protected resources on your server.
```
the tutorial appears to show how to call our management API rather than on your server...

vibronet commented on articles/quickstart/spa/angular-next/02-calling-an-api.md:
```diff
@@ -0,0 +1,186 @@
+
+You can also [protect an API with Auth0](https://auth0.com/docs/microsites/protect-api/protect-api). We offer multiple [Auth0 API quickstarts](https://auth0.com/docs/quickstart/backend) to help you integrate Auth0 with your backend platform.
+
+When you use Auth0 to protect your API, you delegate the authentication process to a centralized service that ensures only approved applications can access private resources. Your Angular application authenticates the user and receives an [access token from Auth0](https://auth0.com/docs/tokens/concepts/access-tokens). The application can then pass that access token to your API as a credential. In ...
```
I would suggest rephrasing so that the description applies to any API protected by Auth0, as opposed to the current formulation that refers only to the custom API. Example (also note small tweaks like authn->authz):

When an API is protected by Auth0, the authorization process is delegated to a centralized service that ensures only approved applications can access private resources. Your Angular application authenticates the user and, in the process, it also receives an access token from Auth0. The application can then pass that access token to the API as a credential, as prescribed by OAuth2. In turn, the API can use Auth0 libraries (or any libraries capable to process JWT, the open standard Auth0 uses to encode access tokens) to verify the access tokens it receives from the calling application and issue a response with the desired data.

lbalmaceda replied:
Separate note: if we leave "angular" aside, this paragraph will definitely match what we already have for other API articles. In fact, I think you've based yours in the [auth0-react](https://auth0.com/docs/quickstart/spa/react/02-calling-an-api) one. If we could rephrase this to make it platform agnostic (e.g. "your angular app" -> "your app"), we can reuse (docs' snippets) all the introductory text between quickstarts and focus on the code snippets or platform notes.

stevehobbsdev replied:
Thanks @vibronet. Yes this was mostly lifted from the React article as I believed that to be mostly correct, aside from the mention of Angular. You're right @lbalmaceda in that we could abstract this out and use in both places 👍

vibronet commented on articles/quickstart/spa/angular-next/02-calling-an-api.md:
```diff
@@ -0,0 +1,186 @@
+
+## Configure Your App to Make Authorized API Calls
+
+To facilitate calling APIs, the Angular SDK exports an [`HttpInterceptor`](https://angular.io/api/common/http/HttpInterceptor) that automatically adds an `Authorization` header with the access token to outgoing requests. However, only requests that are specified through configuration will have this header attached, preventing unintended leakage of access tokens to unexpected URLs.
```
i suggest (emphasis just here, not in the doc)

> the _appropriate_ access token

so that we don;t imply that there's only one

vibronet commented on articles/quickstart/spa/angular-next/02-calling-an-api.md:
```diff
@@ -0,0 +1,186 @@
+
+The `AuthModule` setup is similar to the one discussed in the [Configure the `Auth0Provider` component](/quickstart/spa/angular-next##register-and-configure-the-authentication-module) section: you import the `AuthModule` type from the SDK into your Angular application module and call the static `forRoot` method, passing the `domain` and `clientId` properties. The values of these two props come fr ...
+
+However, your Angular application needs to pass an access token when it calls a target API to access private resources. You can [request an access token](https://auth0.com/docs/tokens/guides/get-access-tokens) in a format that the API can verify by passing the `audience` and `scope` props to `AuthModule.forRoot`.
```
you can request an access token _at user authentication time_

lbalmaceda replied:
nit: isn't "props" an angular term? this repeats across the doc

stevehobbsdev replied:
Perhaps you meant _React_ here but yes - React has more or less coined "props" to mean something specific to React. I'll expand to "properties" throughout.

vibronet commented on articles/quickstart/spa/angular-next/02-calling-an-api.md:
```diff
@@ -0,0 +1,186 @@
+
+In addition, the configuration should specify which requests should have an `Authorization` header added to them with the access token.
+
+The following example shows how to set up your Angular application module, with all the necessary imported types and the configuration to call the management API:
```
Does this overrides stuff that the reader might have written in the former chapter? Is there an alternative path in whihc this code is added to a clean starting point? In both cases, it seems the instructions should be more specific

vibronet commented on articles/quickstart/spa/angular-next/02-calling-an-api.md:
````diff
@@ -0,0 +1,186 @@
+  bootstrap: [AppComponent],
+})
+```
+
````
if there's any way of referencing the section of the preceding chapter where the other config essential entries are explained, would be good to do so- or at least to mention that those explanations are there, and here we focus only on the new, API related stuff

vibronet commented on articles/quickstart/spa/angular-next/02-calling-an-api.md:
```diff
@@ -0,0 +1,186 @@
+
+Notice the `httpInterceptor` configuration, which specifies:
+
+* Any request where the URL starts with `https://${account.namespace}/api/v2/` should have an access token attached
```
longer term comment.
hmm, `account.namespace` looks odd... I am sure you are using this to be consistent with what we have done elsewhere: given that in other SDKs we use different terms (domain? tenant?) it might be worth to do a wide sweep to track what we use across the board, and begin a consolidation effort

lbalmaceda replied:
`${account.namespace}` is the template string that the docs app will later process and replace with the "AUTH0_DOMAIN" or the actual app' domain if they are logged in.
Ref: https://github.com/auth0/docs/blob/master/CONTRIBUTING.md#user-specific-variables

Given that in this article Steve explains how to use the mgmt API, the audience must match the app's tenant.

vibronet commented on articles/quickstart/spa/angular-next/02-calling-an-api.md:
```diff
@@ -0,0 +1,186 @@
+:::note
+In the case of the Auth0 Management API, the `read:current_user` and `update:current_user_metadata` scopes let you get an access token that can retrieve user details and update the user's information. In the case of your APIs, you'll define custom [API scopes](https://auth0.com/docs/scopes/current/api-scopes) to implement access control, and you'll identify them in the calls that your client appl ...
+:::
+
```
All the info above is useful, but I don;t know if it belongs to a quickstart... still feel a lot like docs to me. We need to have that info somewhere, but having it here seems to make the text longer.
Making the working assumption that we keep the current "doc" approach: this would be a good place to explain that all the settings described affect the AT you get at sign in time; if they want a different token for another API, or new scopes, they need to do something else (and hint at what)

stevehobbsdev replied:
Agreed, I'm happy to remove or change it. This is also present in the React quickstart, which I was using as a known good base.

vibronet commented on articles/quickstart/spa/angular-next/02-calling-an-api.md:
```diff
@@ -0,0 +1,186 @@
+
+## Make an API Call
+
+With your app module configured with the HTTP interceptor from the Angular SDK, calls you make to the Auth0 Management API will have a valid access token specified in the `Authorization` header. Let's use this as an example for showing user metadata.
```
we can't know if it's valid, the API will decide that :) we can only know whether it's expired or not. We should briefly mention that we also refresh tokens automatically before use, when necessary, otherwise some developers will try to do it themselves and dig around for the RT

lbalmaceda replied:
I agree. Maybe we could use the "correct" access token instead.

lbalmaceda commented on articles/quickstart/spa/angular-next/01-login.md:
```diff
@@ -107,17 +107,16 @@ import { AuthService } from '@auth0/auth0-angular';
-      *ngIf="auth.isAuthenticated$ | async"
-      (click)="auth.logout({ returnTo: 'http://localhost:4200' })">
+    <ng-container *ngIf="auth.isAuthenticated$ | async; else loggedOut">
+      <button (click)="auth.logout({ returnTo: 'http://localhost:4200' })">
```
will the user remember this URL should target their app's base URL?

stevehobbsdev replied:
I've modified this to get it generically rather than encode the URL, should cure that issue 👍

lbalmaceda commented on articles/quickstart/spa/angular-next/02-calling-an-api.md:
```diff
@@ -0,0 +1,186 @@
+  - angular
+  - api
+github:
+  path: Sample-01
```
haven't checked, but the same sample app as the first quickstart? 

```suggestion
  path: Sample-01
```

stevehobbsdev replied:
Yes - both articles link to the same sample, which demonstrates all the use cases together.

lbalmaceda commented on articles/quickstart/spa/angular-next/02-calling-an-api.md:
```diff
@@ -0,0 +1,186 @@
+
+### Configure the Auth Module
+
+The `AuthModule` setup is similar to the one discussed in the [Configure the `Auth0Provider` component](/quickstart/spa/angular-next##register-and-configure-the-authentication-module) section: you import the `AuthModule` type from the SDK into your Angular application module and call the static `forRoot` method, passing the `domain` and `clientId` properties. The values of these two props come fr ...
```
I couldn't find this section above. Is this linking to another page? Also, is this double # on the link correct?
```suggestion
The `AuthModule` setup is similar to the one discussed in the [Configure the `Auth0Provider` component](/quickstart/spa/angular-next#register-and-configure-the-authentication-module) section: you import the `AuthModule` type from the SDK into your Angular application module and call the static `forRoot` method, passing the `domain` and `clientId` properties. The values of these two props come from the ["Settings" values](https://auth0.com/docs/quickstart/spa/angular-next#configure-auth0) of the single-page application you've registered with Auth0.
```

stevehobbsdev replied:
Yes it's linking to the first tutorial - I'll make the link more explicit.

I've applied the changes locally, nice spot on the double hash 👍

lbalmaceda commented on articles/quickstart/spa/angular-next/02-calling-an-api.md:
```diff
@@ -0,0 +1,186 @@
+* Any request where the URL starts with `https://${account.namespace}/api/v2/` should have an access token attached
+* If the URL is matched, the attached access token should have an audience value of `https://${account.namespace}/api/v2/` and the `read:current_user` scope;
+
+A more detailed run-down of the available configuration options for the HTTP interceptor is available in [the repository docs](https://github.com/auth0/auth0-angular#call-an-api).
```
ideally, we should keep everything in the docs site and not reference the github repo. We already have one for the spa sdk here https://auth0.com/docs/libraries. Consider in the future, adding one for the wrapper SDKs.

stevehobbsdev replied:
Yes, we are planning one. We do have one for React and are planning one for Angular. I'll leave the link until the document is live, but yes - let's link to library docs instead.

lbalmaceda commented on articles/quickstart/spa/angular-next/02-calling-an-api.md:
```diff
@@ -0,0 +1,186 @@
+    .pipe(
+      concatMap((user) =>
+        this.http.get(
+          `https://angular-workshop.us.auth0.com/api/v2/users/<%= "${user.sub}" %>`
```
should this URL use the ${account.namespace} instead? 

also, given we're appending the -sub- value instead of using some utility method to construct the query, do we need to URL encode this? thinking of the pipe char

stevehobbsdev replied:
Yes, good point on both counts. I've removed the absolute URL and used `encodeURI`. The request actually works either way, but encoding it is the right thing to do.

lbalmaceda reviewed (changes requested):
I don't have access to the preview link so I reviewed the .md file. In general looks good, I agree that we shouldn't need to explain in detail every auth0 product feature here. We could briefly mention them if absolutely necessary for context, and link them to external articles. Note that much of this content can be re-used across SPA articles so would be good to generify the paragraphs and just reference to those "snippet files". Also, there are 2 big code snippets (module setup and user profile component) that can be split into 2 or 3 more granular snippets. 

e.g. a first piece with the required option parameters, explaining how the default flow/scenario behaves. A second piece with some default overrides, for the second most common scenarios. And last, the third piece with optional parameters that go over the normal flow.

Commit:
Apply smaller change requests from review

Commit:
Modify API2 preamble for Angular and extract to include file for React

Commit:
Update API tutorial description for both Angular and React

Commit:
Adjust context around module configuration snippet

Commit:
Strip down the SPA API intro

Commit:
Further refine detail on 'call an api'

lbalmaceda commented on articles/quickstart/spa/angular-next/02-calling-an-api.md:
```diff
@@ -0,0 +1,140 @@
+To install and configure the HTTP interceptor, perform the following steps:
+
+* Import the `AuthHttpInterceptor` type from the Auth0 Angular SDK
+* Import `HttpClientModule` and `HTTP_INTERCEPTOR` from `@angular/common/http`
```
haven't checked locally but either this ends with S or the code snippets are wrong
```suggestion
* Import `HttpClientModule` and `HTTP_INTERCEPTORS` from `@angular/common/http`
```

stevehobbsdev replied:
Thanks, have applied this locally.

lbalmaceda commented on articles/quickstart/spa/angular-next/02-calling-an-api.md:
````diff
@@ -0,0 +1,140 @@
+
+The following is an example of an Angular module (based upon the default implementation when you create a new app using `ng new`) that supports `AuthHttpInterceptor`, configured to call the Auth0 Management API with the ability to read the current user profile:
+
+```javascript
````
this is a really big chunk of code. I understand we want to give a lot of context around the lines that we're changing, but consider splitting this snippet into 2 or 3 smaller snippets that go between the steps that you are enumerating above. There's a good visual example of what I mean in the official angular docs https://angular.io/guide/http#provide-the-interceptor

stevehobbsdev replied:
Agreed, I like how the Angular docs do that. Let me break this snippet down in a similar way.

stevehobbsdev replied:
I've broken it down now and pushed the changes - how does it look?

lbalmaceda replied:
Better, thanks!

lbalmaceda commented on articles/quickstart/spa/angular-next/02-calling-an-api.md:
```diff
@@ -0,0 +1,140 @@
+              audience: 'https://${account.namespace}/api/v2/',
+
+              // The attached token should have these scopes
+              scope: 'read:current_user'
```
I understand the base SDK is already enforcing the `openid` scope. The question is, do we want to explicitly request it for giving more visibility that the wrapper is OIDC conformant? Or at least add a comment to imply that these other set of scopes are also being included.

example (have not checked the default scopes)
```js
scope: 'read:current_user' // "openid profile email" are included internally
```

stevehobbsdev replied:
Good question. Here we're talking about scopes for the access token, so is it relevant that the extra default scopes are reference here?

lbalmaceda replied:
I asked out of curiosity. I don't have any preference. 👍

lbalmaceda commented on articles/quickstart/spa/angular-next/02-calling-an-api.md:
```diff
@@ -0,0 +1,140 @@
+
+@Component({
+  selector: 'app-metadata',
+  template: `<div *ngIf="auth.user$ | async as user">
```
the template is not using this `user` variable. I guess is better to either use it explicitly or await with ngIf on `metadata` instead.

stevehobbsdev replied:
Good spot - I've gone with waiting on `metadata`.

Commit:
Reword intro slightly

Commit:
Clarified starting point of the article

Commit:
Changed interceptor heading

Angular uses the terminology "provide" here, as the interceptor is being
added to the `providers` list in the module.

https://angular.io/guide/http#provide-the-interceptor

Commit:
Break down snippet into more manageable chunks

stevehobbsdev commented on articles/quickstart/spa/angular-next/02-calling-an-api.md:
````diff
@@ -0,0 +1,151 @@
+})
+```
+
+Please [refer to the docs](https://github.com/auth0/auth0-angular#configure-authhttpinterceptor-to-attach-access-tokens) for more information on the available options for the HTTP interceptor.
````
We can link to the library docs here when they are published. Right now, I have no idea what that link would be.
